emop-maven-plugin/src/main/java/edu/cornell/VmsMojo.java

```diff
@@ -0,0 +1,277 @@
+                trees.add(new CanonicalTreeParser(null, objectReader, commit.getTree().getId()));
+            }
+        } catch (IOException exception) {
+            throw new MojoExecutionException("Encountered an error when creating trees from commits");
+        }
+
+        // Sets up diffFormatter and analyzes for differences between the two trees
+        diffFormatter.setRepository(git.getRepository());
+        diffFormatter.setContext(0);
+        diffFormatter.setDetectRenames(true);
+        try {
+            diffs = diffFormatter.scan(trees.get(1), trees.get(0));
+        } catch (IOException exception) {
+            throw new MojoExecutionException("Encountered an error when analyzing for differences between commits");
+        }
+
+        git.close();
+        return diffs;
+    }
+
+    /**
+     * Updates the lineChanges and renames based on found differences.
+     *
+     * @param diffs List of differences between two versions of the same program
+     * @throws MojoExecutionException if error is encountered during runtime
+     */
+    private void findLineChangesAndRenames(List<DiffEntry> diffs) throws MojoExecutionException {
+        try {
+            for (DiffEntry diff : diffs) {
+                // If the old path is /dev/null, then the file has been created between commits
+                if (diff.getOldPath().equals("/dev/null")) {
+                    newClasses.add(diff.getNewPath());
+                } else if (!diff.getNewPath().equals("/dev/null")) { // Ignore if a deleted class
+                    // Gets renamed classes
+                    if (!diff.getOldPath().equals(diff.getNewPath())) {
+                        renames.put(diff.getNewPath(), diff.getOldPath());
+                    }
+
+                    // For each file, find the replacements, additions, and deletions at each line
+                    for (Edit edit : diffFormatter.toFileHeader(diff).toEditList()) {
+                        Map<Integer, Integer> lineChange = new HashMap<>();
+                        if (lineChanges.containsKey(diff.getOldPath())) {
+                            lineChange = lineChanges.get(diff.getOldPath());
+                        }
+                        lineChange.put(edit.getBeginA() + 1, edit.getLengthB() - edit.getLengthA());
+                        lineChanges.put(diff.getOldPath(), lineChange);
+                    }
+                }
+            }
+        } catch (IOException exception) {
+            throw new MojoExecutionException("Encountered an error when comparing different files");
+        }
+    }
+
+    /**
+     * Scrubs newViolations of violations believed to be duplicates from violation-counts-old.
+     */
+    private void removeDuplicateViolations() {
+        Set<Violation> violationsToRemove = new HashSet<>();
+        for (Violation newViolation : newViolations) {
+            Set<Violation> relevantOldViolations = oldViolations.stream()
```

Comment:
is this relevantOldViolation always empty or a singleton set? That is, can it ever have more than one match?

Reply:
Yes, if the same violation occurs at least twice in the old file within the offset then they will both be mapped to the same violations in the new file (within the offset)

Reply:
Something I could do is change this filtered set to a for each loop -> if any of the old violations can be successfully mapped to the new one then we can add the new violation to the list of violations to remove, this would save us the cost of iterating through all old violations every time